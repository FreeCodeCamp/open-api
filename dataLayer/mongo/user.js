const validator = require('validator');
const UserModel = require('../model/user.js');
const moment = require('moment');
const fs = require('fs');

import { asyncErrorHandler } from '../../utils';

function doesExist(Model, options) {
  return Model.find(options).exec();
}

export function getUsers(args) {
  return new Promise((resolve, reject) => {
    UserModel.find(args)
      .then(users => {
        let userMap = [];
        users.forEach((user, index) => {
          userMap[index] = user;
        });
        let res = {};
        res.users = userMap;
        resolve(res.users);
      })
      .catch(err => {
        reject(err);
      });
  });
}

export async function createUser(email) {
  const isEmail = validator.isEmail(email);
  if (!isEmail) {
    throw new Error('You must provide a vaild email');
  }
  const entities = await asyncErrorHandler(doesExist(UserModel, { email }));
  const exists = !!entities.length;
  if (exists) {
    throw new Error(`An account with the email ${email} already exists`);
  }
  const newUser = await asyncErrorHandler(
    UserModel.create({ email }),
    'Something went wrong creating your account, please try again'
  );
  return newUser;
}
