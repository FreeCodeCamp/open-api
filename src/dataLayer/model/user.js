const mongoose = require('mongoose');

const Schema = mongoose.Schema;

const userSchema = new Schema({
  accountLinkId: {
    type: 'string',
    description: 'A uuid used to link SSO and freeCodeCamp accounts together',
    required: true
  },
  email: {
    type: 'string'
  },
  newEmail: {
    type: 'string'
  },
  emailVerifyTTL: {
    type: 'date'
  },
  emailVerified: {
    type: 'boolean',
    default: false
  },
  emailAuthLinkTTL: {
    type: 'date'
  },
  progressTimestamps: {
    type: 'array',
    default: []
  },
  isBanned: {
    type: 'boolean',
    description: 'User is banned from posting to camper news',
    default: false
  },
  isCheater: {
    type: 'boolean',
    description:
      'Users who are confirmed to have broken academic honesty policy are ' +
      'marked as cheaters',
    default: false
  },
  githubURL: {
    type: 'string'
  },
  website: {
    type: 'string'
  },
  username: {
    type: 'string'
  },
  bio: {
    type: 'string',
    default: ''
  },
  about: {
    type: 'string',
    default: ''
  },
  name: {
    type: 'string',
    default: ''
  },
  location: {
    type: 'string',
    default: ''
  },
  picture: {
    type: 'string',
    default: ''
  },
  linkedin: {
    type: 'string'
  },
  codepen: {
    type: 'string'
  },
  twitter: {
    type: 'string'
  },
  sendQuincyEmail: {
    type: 'boolean',
    default: true
  },
  isLocked: {
    type: 'boolean',
    description:
      'Campers profile does not show challenges/certificates to the public',
    default: true
  },
  currentChallengeId: {
    type: 'string',
    description: 'The challenge last visited by the user',
    default: ''
  },
  isHonest: {
    type: 'boolean',
    description: 'Camper has signed academic honesty policy',
    default: false
  },
  isFrontEndCert: {
    type: 'boolean',
    description: 'Camper is front end certified',
    default: false
  },
  isDataVisCert: {
    type: 'boolean',
    description: 'Camper is data visualization certified',
    default: false
  },
  isBackEndCert: {
    type: 'boolean',
    description: 'Campers is back end certified',
    default: false
  },
  isFullStackCert: {
    type: 'boolean',
    description: 'Campers is full stack certified',
    default: false
  },
  isRespWebDesignCert: {
    type: 'boolean',
    description: 'Camper is responsive web design certified',
    default: false
  },
  is2018DataVisCert: {
    type: 'boolean',
    description: 'Camper is data visualization certified (2018)',
    default: false
  },
  isFrontEndLibsCert: {
    type: 'boolean',
    description: 'Camper is front end libraries certified',
    default: false
  },
  isJsAlgoDataStructCert: {
    type: 'boolean',
    description:
      'Camper is javascript algorithms and data structures certified',
    default: false
  },
  isApisMicroservicesCert: {
    type: 'boolean',
    description: 'Camper is apis and microservices certified',
    default: false
  },
  isInfosecQaCert: {
    type: 'boolean',
    description:
      'Camper is information security and quality assurance certified',
    default: false
  },
  isChallengeMapMigrated: {
    type: 'boolean',
    description: 'Migrate completedChallenges array to challenge map',
    default: false
  },
  completedChallenges: {
    type: [
      {
        completedDate: 'number',
        numOfAttempts: 'number',
        id: 'string',
        files: 'object',
        bucket: 'string'
      }
    ],
    default: []
  },
  portfolio: {
    type: 'array',
    default: []
  },
  rand: {
    type: 'number',
    index: true
  },
  timezone: {
    type: 'string'
  },
  theme: {
    type: 'string',
    default: 'default'
  },
  languageTag: {
    type: 'string',
    description: 'An IETF language tag',
    default: 'en'
  }
});

module.exports = mongoose.model('User', userSchema);
