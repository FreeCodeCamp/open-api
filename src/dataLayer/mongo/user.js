import validator from 'validator';
import UserModel from '../model/user.js';
import debug from 'debug';

import { asyncErrorHandler } from '../../utils';

const log = debug('fcc:dataLayer:mongo:user');

function doesExist(Model, options) {
  return Model.find(options).exec();
}

export function getUsers(_, args = {}) {
  return new Promise((resolve, reject) => {
    UserModel.find(args)
      .then(users => {
        let userMap = [];
        users.forEach((user, index) => {
          userMap[index] = user;
        });
        let res = {};
        res.users = userMap;
        resolve(res.users);
      })
      .catch(err => {
        reject(err);
      });
  });
}

export async function createUser(email) {
  const isEmail = validator.isEmail(email);
  if (!isEmail) {
    throw new Error('You must provide a vaild email');
  }
  const entities = await asyncErrorHandler(doesExist(UserModel, { email }));
  const exists = !!entities.length;
  if (exists) {
    throw new Error(`An account with the email ${email} already exists`);
  }
  const newUser = await asyncErrorHandler(
    UserModel.create({ email }),
    'Something went wrong creating your account, please try again'
  );
  return newUser;
}
