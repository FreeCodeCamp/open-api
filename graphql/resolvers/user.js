import { createUser, getUsers } from '../../dataLayer/mongo/user';

export const userResolvers = {
  Query: {
    users: getUsers
  },
  Mutation: {
    createUser: (_, { email }) => dbUsers.createUser(email)
  }
};
