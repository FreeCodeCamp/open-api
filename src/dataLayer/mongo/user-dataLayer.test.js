/* global expect beforeAll afterAll */
import { createUser, getUser } from './user';
import { isObject, isEmpty } from 'lodash';
import mongoose from 'mongoose';

const validContext = global.mockedContextWithValidToken;

beforeAll(async function beforeAllTests() {
  await mongoose.connect(global.__MONGO_URI__);
});

afterAll(async function afterAllTests() {
  await mongoose.disconnect();
});

describe('createUser', () => {
  it('should return a User object', done => {
    expect.assertions(2);
    createUser({}, {}, validContext)
      .then(result => {
        const { name, email, accountLinkId } = result;
        // there is some weird Promise thing going on with `result`
        // which screws with lodash.has()
        const hasKeys =
          !isEmpty(name) && !isEmpty(email) && !isEmpty(accountLinkId);

        expect(isObject(result)).toBe(true);
        expect(hasKeys).toBe(true);
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should return user for accountLinkId if found in db', done => {
    expect.assertions(1);
    createUser({}, {}, validContext)
      .then(result => {
        expect(result.accountLinkId).toEqual(
          'a-very-unique-string-for-charlie@thebear.me'
        );
        return;
      })
      .then(done)
      .catch(done);
  });
});

describe('getUser', () => {
  it('should return a User object fo a valid request', done => {
    expect.assertions(2);
    const email = 'charlie@thebear.me';
    getUser({}, { email }, validContext)
      .then(result => {
        const { name, email, accountLinkId } = result;
        // there is some weird Promise thing going on with `result`
        // which screws with lodash.has()
        const hasKeys =
          !isEmpty(name) && !isEmpty(email) && !isEmpty(accountLinkId);

        expect(isObject(result)).toBe(true);
        expect(hasKeys).toBe(true);
        return;
      })
      .then(done)
      .catch(done);
  });

  it('should throw for a user not found', done => {
    expect.assertions(1);
    const email = 'not@inthe.db';
    getUser({}, { email }, validContext).catch(err => {
      expect(err.message).toContain('No user found for ');
      done();
    });
  });

  it('should throw if the supplied email is not valid', done => {
    expect.assertions(15);
    Promise.all([
      getUser({}, { email: 1 }, validContext).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('1');
      }),
      getUser({}, { email: 'not an email' }, validContext).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('not an email');
      }),
      getUser({}, { email: ['nope'] }, validContext).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('["nope"]');
      }),
      getUser({}, { email: { email: false } }, validContext).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('false');
      }),
      getUser({}, { email: 1 }, validContext).catch(err => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err.message).toContain('Expected a valid email');
        expect(err.message).toContain('1');
      })
    ]).then(() => {
      done();
    });
  });
});
