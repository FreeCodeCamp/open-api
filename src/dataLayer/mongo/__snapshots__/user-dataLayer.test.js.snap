// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deleteUser should return with an error message when a user not in db 1`] = `[Error: There is no account with this accountLinkId 85a937d5-c82c-4aa9-8e0b-9f2b9a7cc36c]`;

exports[`deleteUser should return with an error when deleting the user not logged in 1`] = `[Error: You can delete only your account]`;
